README.md
Automatic Turntable Addon
================
Version 0.3.1
-----------
(c) Mark Fitzgibbon 2017

Description
-----------
Automatically focus camera to rotate around a selected object in a scene.
Will default to camera named "Camera", using render settings set for this camera.
Script is provided as-is, under the GNU General Public License.
Details can be located in the attached LICENSE file

Download
--------
User can pull, fork, or download this tool's source code from 
[Blender-Automatic-Turntable Repository](https://github.com/ibbolia/blender-automatic-turntable)

Requirements
------------
Requires Blender, v2.78 preferred.
Code is not intended nor designed to work as a standalone Python program.

Install
-------
User wishing to install as and addon to blender should perform the following:

1. Zip folder (if not already done) 
2. Open Blender program
3. Navigate to File > User Preferences > Add-ons
4. Install from file
5. Select Zip folder containing 

Alternatively, User can install addon files directly in Blender's addon folder
Add-on is only in "Testing" supported level

User is not required to keep hidden git folder to run addon.

Usage
--------
Addon is accessible through Render>Automatic Turntable
Automatic turntable will not manage render settings automaticaly
User should confirm camera parameters (e.g. clipping, FOV) befre execution

User will be presented with a dialog box and the following parameters:
- Iterations: The total amount of renders to be made
- Increments: The angle (in degrees) between iterations
- Margin: Minimum amount of space (percent) between the object and edge of render
- Camera: Name of Camera object to be utilized (defaults to Active)
- File Path: Directory renders will be saved to

User can then press OK to start render process
Note: There is no built in indicator for how much a render batch has progressed

Notes
--------
- [Camera distance formula is dervied from this stackexchange page](http://photo.stackexchange.com/questions/12434/how-do-i-calculate-the-distance-of-an-object-in-a-photo)

Contact
-------
- [Github](https://github.com/ibbolia)
- Twitter: [@ibbolia](https://twitter.com/ibbolia)
- Gmail: mwfitzgibbon

CHANGELOG.md
#Change Log
All notable changes will be tracked in this file

##[Unreleased]
###Changed
- Corrections to Readme
- Corrected Order of Operations on obj_dst lambda (__init__.py, line 55)
- Added reference to origin during Oribital position calculation

###Removed
- Extraneous BoundingBox call (__init__.py, line 152)

##[0.3.1] - 2017-02-05
###Added
- AutomaticTurntableOperator class (To replace Orbital Operator)
- Turntable class (To replace Orbital)
- Keymaps variable
- AutomaticTurntableOperator UI via invoke_props_dialog

###Changed
- Register/Unregister takes effect on AutomaticTurntableOperator class
- Radius calculation now uses camera's focal length and relative size of target
- Filename string formats and uses correct extension string

###Removed
- Active code referencing Orbital class
- Orbital Operator class

##[0.2.3] - 2017-01-16
###Fixed
- Issue 7: Camera rotation points now do not exceed final position
- Issue 8: Utilizes camera defined in script instead of active camera

##[0.2.2] - 2017-01-05
###Fixed
- Issue 5: Transforms no longer negatively affect render fitting

###Changed
- Bounding Box calculation now uses matrices directly on values

##[0.2.1] - 2017-01-04 
###Fixed
- Script is now able to account for global scale values of objects when determining bounds (Issue #5)

###Added
- Significan Known Issues section to readme

###Changed
- Classes seperated into individual files
- Modified specific methodology of corner calculation in BoundingBox.py
- (Above) Added Global Scale to calculation method for an individual bounding box
- Date markings in Readme

###Removed
- addon_menu.py: according to bpy documentation, this is the intent of an init file

##[0.2] - 2016-12-28
###Added
- RadialPoints.py file for determining position of camera during calculations
	-Some code exists for using spherical coordinates. It is not functional.
- addon_menu.py for setting up UI menu
- Altered calculation for bounding box 
	-(based on [boundingbox](https://github.com/ibbolia/blender-scripts/))
	-Calculation still utilizes "fit_view_to_selected" method, to be phased out

###Changed
- Functionality to use number of steps instead of degrees
- Camera radius from origin zoomed back around 10% from previous 
- Readme edits
	- Corrected link to github repository
	- Added link to twitter profile
	- Version number change


##[0.1] - 2016-12-14
###Added
- Base __init__.py file (from blender-scripts)

Orbital.py
import bpy
import mathutils
from math import radians
from . import RadialPoints

class Orbital:
	'''Orbital class contains variables and functions for positioning and rotating camera around a model
		'''
	def __init__(self, box = None):
		self.TT_FILEPATH_ROOT = "/tmp\\"
		self.TT_FILEPATH_ITERATOR = "1"
		self.TT_FILEPATH_EXT  =".png"

		self.TT_ANGLE_INCREMENTS = 90
		self.TT_POSITION = (0.0, -5.0, 0.0)	#Camera should be in front (Neg-Y) of model

		self.TT_ROTATION_X = 90.0
		self.TT_ROTATION_Y = 0.0
		self.TT_ROTATION_Z = 0.0

		self.tt_orbit = RadialPoints.CircularPositioning()
		if box is not None:
			self.tt_orbit.tt_origin = box.midpoint 
		self.tt_iterations = 3

	def setToFrontview(self, camera):
		'''sets camera of choice to a directly front view position (assuming "Front" is theta=270 degrees)
			camera -- Camera object to be positioned
			'''
		#Move camera front and center
		camera.location = self.TT_POSITION
		camera.rotation_euler = (radians(self.TT_ROTATION_X), 0.0, 0.0)

		#reposition camera to fit selected items
		bpy.ops.view3d.camera_to_view_selected()
		camera.location = (camera.location[0], camera.location[1]+ (camera.location[1]*.1), camera.location[2])
		rads = self.tt_orbit.radiusToPoint(camera.location)
		self.tt_orbit.tt_circular_coords = [rads, 270]
		print("Orbiting around: " + str(self.tt_orbit.tt_origin) )

	def catchRender(self, camera):
		bpy.data.scenes['Scene'].render.filepath = self.TT_FILEPATH_ROOT + self.TT_FILEPATH_ITERATOR + self.TT_FILEPATH_EXT
		bpy.ops.render.render( write_still=True ) 
		self.TT_FILEPATH_ITERATOR = str(int(self.TT_FILEPATH_ITERATOR)+1)

	def setCameraToNextInc(self, camera  ):
		sampPos = self.tt_orbit.getPointXYZ()
		
		self.TT_ROTATION_Z = self.TT_ROTATION_Z + self.TT_ANGLE_INCREMENTS
		camera.rotation_euler = (radians(self.TT_ROTATION_X), 0.0, radians(self.TT_ROTATION_Z))
		
		#Set camera position
		self.tt_orbit.addToAngle(self.TT_ANGLE_INCREMENTS)
		sampPos = self.tt_orbit.getPointXYZ()
		camera.location= mathutils.Vector(sampPos)

	def renderOrbit(self, camera):
		'''Iterate tghrough radial steps and render at position
			camera -- Camera object to perform on
			'''
		#Removes first render jitter
		sampPos = self.tt_orbit.getPointXYZ()
		camera.location= mathutils.Vector(sampPos)

		while(self.tt_iterations >0):
			#Take render
			#print("Iteration #" + str(self.tt_iterations))
			self.catchRender(camera)
			self.tt_iterations -= 1
			if(self.tt_iterations is not 0):
				#Rotate camera
				self.setCameraToNextInc(camera)
			
	

BoundingBox.py
from mathutils import Vector
import math

class BBox:
	def __init__(self, obj):
		xMax = -math.inf
		yMax = -math.inf
		zMax = -math.inf

		xMin = math.inf
		yMin = math.inf
		zMin = math.inf

		self.mx = obj.matrix_world

		for b in obj.bound_box:
			xMax = max(xMax, b[0])
			yMax = max(yMax, b[1])
			zMax = max(zMax, b[2])

			xMin = min(xMin, b[0])
			yMin = min(yMin, b[1])
			zMin = min(zMin, b[2])

		self.mini = self.mx * Vector((xMin, yMin, zMin))
		print("Min: " + str(self.mini))
		self.maxi = self.mx * Vector((xMax, yMax, zMax))
		print("Max: " + str(self.maxi))

class BoundingBox:
	def __init__(self, lst=[]):
		xMax = -math.inf
		yMax = -math.inf
		zMax = -math.inf

		xMin = math.inf
		yMin = math.inf
		zMin = math.inf

		boxen = []
		#Obtain bounding box in an object's local space
		for ob in lst:
			boxen.append(BBox(ob))

		for b in boxen:
			xMax = max(xMax, b.maxi[0])
			yMax = max(yMax, b.maxi[1])
			zMax = max(zMax, b.maxi[2])

			xMin = min(xMin, b.mini[0])
			yMin = min(yMin, b.mini[1])
			zMin = min(zMin, b.mini[2])

		xMid = (xMin + xMax)/2
		yMid = (yMin + yMax)/2
		zMid = (zMin + zMax)/2

		self.minimum = (xMin, yMin, zMin)
		self.maximum = (xMax, yMax, zMax)
		self.midpoint = (xMid, yMid, zMid)
		print(self.midpoint)

	def getAxisLength(self, ax):
		return self.maximum[ax] - self.minimum[ax]

RadialPoints.py
import math
from math import radians

class CircularPositioning:
	def __init__(self, orig = (0.0, 0.0, 0.0)):
		self.tt_circular_coords = [0.0, 0.0] 
		self.tt_origin = orig

	def radiusToPoint(self, xyzTuple = (0, 0, 0)):
		if (self.tt_origin == xyzTuple):
			#Radius is zero
			return 0
		#r^2 = (x-h)^2 + (y-j)^2
		rSquared = math.pow((xyzTuple[0] - self.tt_origin[0]), 2) + math.pow((xyzTuple[1] - self.tt_origin[1]), 2)
		return math.sqrt(rSquared)

	def getRadius(self):
		return self.tt_circular_coords[0]

	def getPointXYZ(self):
		xPos = self.tt_circular_coords[0] * math.cos(radians(self.tt_circular_coords[1])) + self.tt_origin[0]
		yPos = self.tt_circular_coords[0] * math.sin(radians(self.tt_circular_coords[1])) + self.tt_origin[1]
		zPos = self.tt_origin[2]
		return (xPos, yPos, zPos)

	def addToAngle(self, theta):
		angle = self.tt_circular_coords[1] + theta

		while angle >= 360:
			angle = angle - 360
		while angle < 0:
			angle = angle + 360
		self.tt_circular_coords[1] =  angle

	def angleToPoint(self, xyzTuple = (0,0,0)):
		pass



class SphericalPositioning:
	#Framework for future improvements
	def __init__(self):
		self.tt_spherical_coords = (0.0, 0.0, 0.0) 
		self.tt_origin = (0.0, 0.0, 0.0)

	def radiusToPoint(self, xyzTuple = (0,0,0)):
		if (self.tt_origin == xyzTuple):
			#Radius is zero
			return 0
		rSquared = math.pow((xyzTuple[0] - self.tt_origin[0]), 2) + math.pow((xyzTuple[1] - self.tt_origin[1]), 2)+ math.pow((xyzTuple[2] - self.tt_origin[2]), 2)
		return math.sqrt(rSquared)



__init__.py
# BEGIN GPL BLOCK    
#    Batch renders around selection to create turntable frames
#    Copyright (C) 2017  Mark Fitzgibbon
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# END GPL BLOCK
bl_info = {
	"name":"Automatic Turntable",
	"description":"Automatically focus camera to rotate around a selected object in a scene",
	"version":(0,3,1),
	"blender":(2,78,0),
	"support":"TESTING",
	"category":"Render",
	"location":"Render",
	"author":"Mark Fitzgibbon"
}
#Python imports
import mathutils
from math import radians

#Blender Imports
import bpy
from bpy.props import IntProperty, FloatProperty, StringProperty

#Local imports
from . import BoundingBox
from . import RadialPoints

addons_keymap = []

def fitCameraToBox(camera, box, margin = 0.0):
	data_cam = bpy.data.cameras[camera.name]
	focal = data_cam.lens

	#Get X, Y, and Z bounding box lengths	
	box_width_x = box.getAxisLength(0)
	box_width_y = box.getAxisLength(1)
	box_height = box.getAxisLength(2)
	
	sensorWidth = data_cam.sensor_width
	sensorHeight = data_cam.sensor_height

	obj_dst = lambda n, m: ( ((focal * n*1000 )/m)/1000 * (1+margin)) + max(box_width_x, box_width_y)/2
	
	#Determine necessary radii for fitting the whole image in
	cam_radius_h = obj_dst(box_height, sensorHeight) 
	cam_radius_w = obj_dst(max(box_width_x, box_width_y), sensorWidth) 

	##Return the furthest (and best fitting) radius
	return max(cam_radius_h, cam_radius_w)

class Turntable:

	def __init__(self, camera, filepath="//render_out",iterations=1, increments=90 ):
		self.iterations=iterations
		self.increments=increments
		self.camera=camera	#Object of camera (eg bpy.data.objects)
		self.filepath=filepath
		self.camera_rotation = (90, 0, 0)

		#Set target center
		self.box = BoundingBox.BoundingBox(bpy.context.selected_objects)
		self.orbit = RadialPoints.CircularPositioning(self.box.midpoint)

		#Set Radius and starting angle
		self.orbit.tt_circular_coords = [fitCameraToBox(self.camera, self.box, 0), 270]	#Angle 270 should translate to a Front view

		#List of all camera position/rotation comos in order
		self.camera_atlas = []
		flag = self.iterations
		while(flag > 0):
			posit = self.orbit.getPointXYZ()	#Location of camera
			rotat = (radians(self.camera_rotation[0]), radians(self.camera_rotation[1]), radians(self.camera_rotation[2]))
			self.camera_atlas.append([posit, rotat])
			self.orbit.addToAngle(self.increments)
			self.camera_rotation = (self.camera_rotation[0], self.camera_rotation[1], self.camera_rotation[2] + increments)
			flag -= 1

	def initializeCamera(self):
		#Activate requested camera if it isn't
		if self.camera is not bpy.context.scene.camera:
			bpy.context.scene.camera = self.camera

		bpy.context.scene.camera.location[2] = self.box.midpoint[2]

	def setToIndexAndRender(self, index = 0):
		#Set position and rotation of camera to position on list at index
		self.camera.location = mathutils.Vector(self.camera_atlas[index][0])
		self.camera.rotation_euler = self.camera_atlas[index][1]

		bpy.data.scenes[bpy.context.scene.name].render.filepath = self.filepath + format(index+1, '03d')
		bpy.ops.render.render( write_still=True )

	def renderAllPositions(self):
		self.initializeCamera()
		index = 0
		for cpos in self.camera_atlas:
			self.setToIndexAndRender(index)
			index += 1

class AutomaticTurntableOperator(bpy.types.Operator):
	# Interface with blender API to use Automatic Turntable
	bl_idname = "render.automatic_turntable"
	bl_label = "Automatic Turntable Renders"
	bl_options = {'REGISTER', 'UNDO'}

	#Number of renders to take
	iterations = IntProperty(
		name= "Iterations",
		min = 1,
		default = 1)
	#Angle between iterations
	increments = FloatProperty(
		name ="Increments",
		min = 0,
		max = 360,
		default = 90
		)
	#Space between most extreme edge and edge of image
	margin = FloatProperty(
		name = "Margin",
		min = 0,
		max = 1,
		default = 0
		)
	#String name of camera
	camera = StringProperty(
		name="Camera"
		)
	#Filepath locaton to save renders
	filepath = StringProperty(
		name="File Path")

	def invoke(self, context, event):
		# Call dialog box 
		actScene = bpy.context.scene
		self.camera =actScene.camera.name
		self.filepath = bpy.data.scenes[actScene.name].render.filepath
		return context.window_manager.invoke_props_dialog(self)

	def execute(self, context):
		ttb=Turntable(bpy.data.objects[self.camera], self.filepath, self.iterations, self.increments)
		ttb.renderAllPositions()
		bpy.data.scenes[bpy.context.scene.name].render.filepath =  self.filepath
		return {'FINISHED'}

def addToRenderMenu(self, context):
	self.layout.operator(
		AutomaticTurntableOperator.bl_idname,
		text = "Automatic Turntable"
		)

def register():
	bpy.utils.register_class(AutomaticTurntableOperator)
	bpy.types.INFO_MT_render.append(addToRenderMenu)

def unregister():
	bpy.utils.unregister_class(AutomaticTurntableOperator)
	bpy.types.INFO_MT_render.remove(addToRenderMenu)

if __name__ == "__main__":
	register()